civismlext/hyperband.py

```diff
@@ -298,45 +299,78 @@ def fit(self, X, y=None, groups=None, **fit_params):
-                            return_times=True, return_parameters=True,
-                            error_score=self.error_score))
-
-                _out = Parallel(
-                    n_jobs=self.n_jobs, verbose=self.verbose,
-                    pre_dispatch=pre_dispatch)(_jobs)
-
-                results, _ = self._process_outputs(_out, n_splits)
+            ns.append(int(np.ceil(B / R * np.power(self.eta, s) / (s + 1.0))))
+            rs.append(int(R / np.power(self.eta, s)))
+            Ts.append(list(ParameterSampler(
+                self.param_distributions,
+                ns[-1],
+                random_state=self._random_state)))
+        nums = copy.copy(ns)
+        offsets = [0] + list(
+            np.cumsum(np.array(nums) * n_splits).astype(int))[:-1]
+
+        for rnd in range(0, smax + 1):
+            # set the costs for this round
+            r_rnd = []
+            for ind, s in enumerate(range(smax, -1, -1)):
+                _r = int(rs[ind] * np.power(self.eta, rnd))
+                r_rnd += [_r] * nums[ind]
+
+            # run the jobs
+            _jobs = []
+            for parameters, _r in zip(
+                    itertools.chain.from_iterable(Ts), r_rnd):
+                _parameters = copy.deepcopy(parameters)
+                _parameters.update(
+                    {list(self.cost_parameter_max.keys())[0]: _r})
+                for train, test in cv_iter:
+                    _jobs.append(delayed(_fit_and_score)(
+                        clone(base_estimator), X, y, self.scorer_,
+                        train, test, self.verbose, _parameters,
+                        fit_params=fit_params,
+                        return_train_score=self.return_train_score,
+                        return_n_test_samples=True,
+                        return_times=True, return_parameters=True,
+                        error_score=self.error_score))
+            _out = Parallel(
+                n_jobs=self.n_jobs, verbose=self.verbose,
+                pre_dispatch=pre_dispatch)(_jobs)
+            out += _out
+
+            # now post-process
+            new_Ts = []
+            new_nums = []
+            for ind, s in enumerate(range(smax, -1, -1)):
+                n_i = int(np.floor(ns[ind] / np.power(self.eta, rnd)))
                 num_to_keep = int(np.floor(n_i / self.eta))
-                sind = np.argsort(results["rank_test_score"])
-                msk = np.zeros(len(results['rank_test_score']))
-                msk[sind[0:num_to_keep]] = 1
-                msk = msk.astype(bool)
-                T = [p for k, p in enumerate(results['params']) if msk[k]]
-
-                out += _out
+                # keep for next round only if num_to_keep > 0 AND
+                # the round after this round would have been executed
+                if num_to_keep > 0 and rnd < s:
+                    _out_s = _out[
+                        offsets[ind]:offsets[ind] + nums[ind] * n_splits]
```

Comment:
Can you add parentheses around the second value here, to clarify the operator precedence? Like this: `_out[offsets[ind]:(offsets[ind] + nums[ind] * n_splits)]`